File: php-checks/src/test/resources/checks/security/POSIXFilePermissionsCheck.php
<?php

namespace N1;

$foo = "0007";
chmod("foo", 0777);       // Noncompliant
chmod("foo", $foo);       // Noncompliant
chmod("foo", 750);        // Noncompliant
chmod("foo", "750");      // Noncompliant
chmod("foo", "0750");     // Noncompliant
Chmod("foo", "0750");     // Noncompliant
\Chmod("foo", "0750");    // Noncompliant
chmod(permissions:0777, filename:"foo");       // Noncompliant

chmod("foo", 0750);       // Compliant
chmod("foo", 0);          // Compliant
chmod("foo", "0120");     // Compliant
chmod("foo", $mode);      // Compliant (unknown)
chmod(750);               // Compliant
chmod(0777, "foo");       // Compliant

$bar = 0750;
umask(0000);              // Noncompliant
umask(0750);              // Noncompliant
umask($bar);              // Noncompliant
umask(0023);              // Noncompliant
Umask(0023);              // Noncompliant
Umask(mask:0023);         // Noncompliant

umask("0023");            // Compliant
umask(0777);              // Compliant
umask(0007);              // Compliant
umask($mask);             // Compliant (unknown)
umask();                  // Compliant

namespace Symfony;

use Symfony\Component\Filesystem\Filesystem;

$foo = 0007;
$fs = new Filesystem();
$fs->chmod("foo", 0777, 0000);   // Noncompliant
$fs->chmod("foo", 0777);         // Noncompliant
$fs->chmod("foo", $foo);         // Noncompliant
$fs->chmod("foo", "0750");       // Noncompliant
$fs->chmod(mode:"0750", "foo");                // Noncompliant
$fs->chmod(umask:0, mode:"0750", "foo");       // Noncompliant

$fs->chmod("foo", 0750);         // Compliant
$fs->chmod("foo", 0);            // Compliant
$fs->chmod("foo", "0120");       // Compliant
$fs->chmod("foo", 0000, 0777);   // Compliant
$fs->chmod("foo", 0777, 0777);   // Compliant
$fs->foo("foo", 0777);           // Compliant
$fs->chmod(0777);                // Compliant
$fs->foo(0777, 0000, "foo");     // Compliant

File: php-checks/src/main/java/org/sonar/php/checks/security/POSIXFilePermissionsCheck.java
/*
 * SonarQube PHP Plugin
 * Copyright (C) 2010-2025 SonarSource Sàrl
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the Sonar Source-Available License Version 1, as published by SonarSource SA.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the Sonar Source-Available License for more details.
 *
 * You should have received a copy of the Sonar Source-Available License
 * along with this program; if not, see https://sonarsource.com/license/ssal/
 */
package org.sonar.php.checks.security;

import java.util.Optional;
import org.sonar.check.Rule;
import org.sonar.php.checks.utils.CheckUtils;
import org.sonar.plugins.php.api.tree.Tree.Kind;
import org.sonar.plugins.php.api.tree.declaration.CallArgumentTree;
import org.sonar.plugins.php.api.tree.expression.ExpressionTree;
import org.sonar.plugins.php.api.tree.expression.FunctionCallTree;
import org.sonar.plugins.php.api.tree.expression.LiteralTree;
import org.sonar.plugins.php.api.visitors.PHPVisitorCheck;

@Rule(key = "S2612")
public class POSIXFilePermissionsCheck extends PHPVisitorCheck {

  private static final String MESSAGE = "Make sure this permission is safe.";

  @Override
  public void visitFunctionCall(FunctionCallTree tree) {
    String functionName = CheckUtils.lowerCaseFunctionName(tree);
    if (tree.callee().is(Kind.OBJECT_MEMBER_ACCESS)) {
      if ("chmod".equals(functionName)) {
        chmodSymfonyAndLaravelCheck(tree);
      }
    } else if ("chmod".equals(functionName)) {
      chmodCoreCheck(tree);
    } else if ("umask".equals(functionName)) {
      umaskCheck(tree);
    }
    super.visitFunctionCall(tree);
  }

  private void chmodCoreCheck(FunctionCallTree tree) {
    Optional<CallArgumentTree> permissionsArgument = CheckUtils.argument(tree, "permissions", 1);
    int mode = permissionsArgument.isPresent() ? resolveArgument(permissionsArgument.get(), 0) : 0;
    if (mode % 8 != 0) {
      context().newIssue(this, tree, MESSAGE);
    }
  }

  private void chmodSymfonyAndLaravelCheck(FunctionCallTree tree) {
    Optional<CallArgumentTree> modeArgument = CheckUtils.argument(tree, "mode", 1);
    Optional<CallArgumentTree> umaskArgument = CheckUtils.argument(tree, "umask", 2);
    int mode = modeArgument.isPresent() ? resolveArgument(modeArgument.get(), 0) : 0;
    int umask = umaskArgument.isPresent() ? resolveArgument(umaskArgument.get(), 0) : 0;

    if ((mode & ~umask) % 8 != 0) {
      context().newIssue(this, tree, MESSAGE);
    }
  }

  private void umaskCheck(FunctionCallTree tree) {
    Optional<CallArgumentTree> maskArgument = CheckUtils.argument(tree, "mask", 0);
    int mask = maskArgument.isPresent() ? resolveArgument(maskArgument.get(), 7) : 7;
    if (mask % 8 != 7) {
      context().newIssue(this, tree, MESSAGE);
    }
  }

  private int resolveArgument(CallArgumentTree argument, int defaultValue) {
    ExpressionTree argumentExpressionTree = CheckUtils.assignedValue(argument.value());
    if (argumentExpressionTree.is(Kind.REGULAR_STRING_LITERAL, Kind.NUMERIC_LITERAL)) {
      String literal = ((LiteralTree) argumentExpressionTree).value();
      return getDecimalRepresentation(literal, defaultValue);
    }
    return defaultValue;
  }

  private static int getDecimalRepresentation(String argument, int defaultValue) {
    if (argument.matches("\"[0-9]*\"")) {
      return Integer.valueOf(CheckUtils.trimQuotes(argument));
    } else if (argument.matches("^0[0-7]*$")) {
      return Integer.parseInt(argument, 8);
    }
    try {
      return Integer.parseInt(argument);
    } catch (NumberFormatException e) {
      return defaultValue;
    }
  }

}
